### ISSUE TITLE
Task keep running after cancel

### ISSUE THREAD

**Issue by zefirka** [redux-saga/redux-saga#1879]
### Steps to reproduce

We have 2 sagas: 

```js
function* pollNewDocuments() {
    while (true) {
        yield delay(DOCUMENTS_POLLING_TIMEOUT);
        yield call(pullDocuments);
    }
}

function* initDocumentsPage() {
    let pollingTask = yield fork(pollNewDocuments);

    yield takeEvery(searchDocuments, function* onSearchDocuments({payload}) {
        if (payload) {
            yield cancel(pollingTask);
            yield call(searchDocumentsSaga, payload.query);
        } else {
            yield call(dropSearchDocuments, docs);
            pollingTask = yield fork(pollNewDocuments);
        }
    });
}
```

#### Description of the bug/issue

The point is that we have 2 sagas. First one (`pollNewDocuments`) just fetching data every `DOCUMENTS_POLLING_TIMEOUT` milliseconds, the second one `initDocumentsPage` forks when we visit documents page. And when we receive `searchDocuments` action we canceling polling task in case of search query is not empty. But sometimes `pollNewDocuments` task doesn't canceling and it still keeps calling `pullDocuments`. Moreover when I tried wrap task into `try { } finally { console.log(yield cancelled()) }` it returns true, but `while(true)` loop keeps running.  

### Environment information
 - `redux-sag": "^0.16.0"`
 - `babel: "^6.26.0"`

**Comment by Andarist** [comment]
I suspect there is some (maybe subtle) bug in your code. Ofc can't be sure but to help you you would have to provide a codesandbox with the issue reproduced on it so I could take a look.

**Comment by zefirka** [comment]
After research I've found that `searchDocuments` action could be dispatched twice with empty payload which forking `pollNewDocuments` and rewriting already running task. So we loose context to running task and could not cancel it. So decision was into: 

```js
  if (pollingTask.isCancelled()) {
    pollingTask = yield fork(pollNewDocuments);
  }
```

Excuse me for false alarm. Thanks for attention and be careful with your sagas. Closing this issue.
